### PR TITLE
Add initial PromiseRegistry implementation

### PR DESCRIPTION
**Pull request by matthewrfennell**
Initial draft PR for [this comment](https://github.com/monal-im/Monal/pull/1183#issuecomment-2378317049). Sorry it's taken so long!

Todolist:

- [x] Properly test in bad network conditions - swipe app away before reply from backend is received
Conclusion: it doesn't help if the connection/stream gets closed, as on resuming the stream Monal will still be using the old password, and the connection will fail.
To fix this, we could store a set of passwords in the keychain and try each one until we find the vaild one (and mark it as the only password).
- [x] Persist and load the promise registry UUIDS `realPersistState` and `realReadState`, so that the extension can mark promises as handled.
- [ ] Properly handle errors from the server - add a parameter in the handlers to get the reply

### CODE REVIEW

File: Monal/Classes/MLPromiseRegistry.m
Comment (tmolitor-stud-tu) [inline Monal/Classes/MLPromiseRegistry.m]: you can use the `nilExtractor()` from `MLConstants.h` here. There is a corresponding `nilWrapper()`, too.

File: Monal/Classes/MLPromiseRegistry.m
Comment (tmolitor-stud-tu) [inline Monal/Classes/MLPromiseRegistry.m]: an `@synchronized` block is the only case where we put the opening `{` into the same line (don't ask me why :D) )

File: Monal/Classes/MLPromiseRegistry.m
Comment (tmolitor-stud-tu) [inline Monal/Classes/MLPromiseRegistry.m]: I sense a terminology confusion here. I'll give a rough introduction into promise terminology:
freshly created promises are in state `pending` (or `unresolved`). that state changes to `resolved` once the promise is `fulfilled` or `rejected`, but never both.
A promise can now be `consumed` by one or more consumers using `then()` or `catch()`.
The closure (called `block` in objc) provided to `then()` will be called with the value given when `fulfilling` the promise, if the promise was `rejected` the `catch()` closure will be called and provided the value given when the promise was `rejected`.

Now back to the original issue: if you mean `consumed` when using `handled`, then that's wrong: promises are for ui only --> the can only be `consumed` in the main app. But the can be `resolved` in the app extension, too.

**Comment by tmolitor-stud-tu** [comment]
I'd suggest a different api feeling more naturally:

A new `MLPromise` class (can well be defined in the same file as the promise registry itself).
That class has an internal member variable holding the uuid of the promise (randomly created on init and directly registering this newly created instance to the promise registry using this uuid). And two other ivars initially being nil, and later holding either the `fulfill` or the `reject` values (see my terminology explanation in the code review).
This class implements the `NSSecureCoding` interface to serialize all these 3 variables when serializing an object of this class.
The class provides 2 methods called `fullfill` and `reject` taking an NSObject as single argument and saving it to either of the two member variables (but doing an NSAssert if trying to fulfill a rejected promise or vice versa).
This class also implements a third method called `toPromise` which will return a promisekit promise (`AnyPromise`). that promise should be `resolved` if either the fulfill or reject ivars are already set when this method is called and a `pending` promise otherwise. that promise should be stored inside an ivar, too, so that calling this method will always return the same promise. the `fullfill` and `reject` methods should use this already created promise, to `resolve` it once called.
The fourth method `_retry` should be used internally by monal after wakeup of the main app from suspension. it should check if an internal `AnyPromise` is present, and `fulfill` or `reject` it based on the two ivars mentioned above. it should always unregister the `MLPromise` from our registry afterwards.

objects of this new class can then be added to the promise registry. the promise registry implements a method called `getPromiseForUuid:` returning such a `MLPromise` object.

how would that be used in code?
- the `-(AnyPromise*) changePassword:(NSString*) newPass` implementation inside `xmpp.m` will create a new `MLPromise` object, call `toPromise` on it and return that (`AnyPromise`).
- the `MLHandler` and its invalidation specified when calling `sendIq` in that `changePassword:` method, will use `getPromiseForUuid:` of the promise registry, to "extract" a `MLPromise` object and call `fulfill` or `reject` on that.
- calling `fulfill` or `reject` while being in the main app (`HelperTools` method `isAppExtension` will return `NO`) should create or use an already existing `AnyPromise` internal to `MLPromise` (see above) and directly fulfill or reject it. it should remove that MLPromise instance from the promise registry afterwards.
- calling `fulfill` or `reject`  while being in an app extension should store the argument in the ivars and don't resolve any promises.
- waking up the main app should reload the serialized promise registry from db (and therefore the serialized `MLPromise` instances and call `_retry` on every `MLPromise` instance inside.

Last but not least, we need a way to provide the error iq to the promise's catch handler. The `AnyPromise` implementation only calls `catch()` if `resolved` with an instance of `NSError`. We should therefore create a category or subclass of `NSError`, somehow adding `XMPPIQ` into it. Maybe use multiple inheritance to create an `XMPPErrorIQ` class inheriting from both `NSError` and `XMPPIQ`. That would overwrite the error description method of `NSError` to use the `HelperTools` method `extractXMPPError:withDescription:` to get the error text.

**Comment by tmolitor-stud-tu** [comment]
We could add a method `getNewPromise` to the promise registry that creates a new `MLPromise` object (and maybe registers it?). That would hide away the `MLPromise` implementation.

Even better, the API described above could be specced in an `Interface` called `MLPromise` (defined in the promise registry header file) and the concrete implementation could be called `MLSerializablePromise` and neatly tucked away inside the promise registry source file (only publicly exposing the `MLPromise` interface).

Anyhow, if we go without that interface, I like the `MLSerializablePromise` name more than just the `MLPromise` one.

**Review by tmolitor-stud-tu — CHANGES_REQUESTED** [review]

File: Monal/Classes/HelperTools.m
Comment (matthewrfennell) [inline Monal/Classes/HelperTools.m:916]: I plan to clean these up, but want to make sure the overall approach is looking better first :)

File: Monal/Classes/MLIQProcessor.m
Comment (matthewrfennell) [inline Monal/Classes/MLIQProcessor.m:799]: I tried splitting `resolve` into `fulfill` and `reject`, but I ended up with two functions and variables in `MLPromise` that seemed to do mostly the same thing. Do you think that just having `resolve` and passing the error is explicit enough? Happy to do it either way, but this way seemed more natural to me.

File: Monal/Classes/MLPromiseRegistry.m
Comment (matthewrfennell) [inline Monal/Classes/MLPromiseRegistry.m:69]: I plan to clean up these `init` methods as there is a lot of duplication between them

File: Monal/Classes/MLIQProcessor.m
Comment (matthewrfennell) [inline Monal/Classes/MLIQProcessor.m:797]: I haven't worked out how to expose the response IQ to the handler yet, but I plan to use it here when I work that out :)

Comment (tmolitor-stud-tu) [reply Monal/Classes/MLIQProcessor.m:797]: I think this could help:

> We should therefore create a category or subclass of NSError, somehow adding XMPPIQ into it. Maybe use multiple inheritance to create an XMPPErrorIQ class inheriting from both NSError and XMPPIQ. That would overwrite the error description method of NSError to use the HelperTools method extractXMPPError:withDescription: to get the error text.

File: Monal/Classes/MLIQProcessor.m
Comment (tmolitor-stud-tu) [reply Monal/Classes/MLIQProcessor.m:799]: Well, that's because the AnyPromise API is overloaded to accept any object and automatically transitions the state to "rejected", if its an instance of NSError (and "fulfilled" if its something different).

Honestly I don't like that design and the Swift part of PromiseKit (the promise lib we are using) has dedicated methods for rejecting and fulfilling.

If, in the long term, the majority of promises used in the objc codebase will be MLPromise objects, I'd like to provide dedicated methods for fulfill and reject (possibly adding an assertion to check if the data provided to reject is an instance of NSError). You are right, internally we could still use only one ivar to hold this data regardless if it was provided calling the fulfill or reject methods.

File: Monal/Classes/MLIQProcessor.m
Comment (tmolitor-stud-tu) [reply Monal/Classes/MLIQProcessor.m:797]: Ah well, now I understand!

You're confused what the invalidation handler is for.

For IQs the normal handler is called for `result` AND `error` IQs. The invalidation handler is only called, if the xmpp session is cleaned up or the IQ timed out (e.g. the result/error IQ will never be received).

I picked that design to allow other parts of our code to clean up in case an IQ result/error is missing.

File: Monal/Classes/MLIQProcessor.m
Comment (tmolitor-stud-tu) [reply Monal/Classes/MLIQProcessor.m:799]: Hmmm, I think we could improve that API further.
What about abstracting the uuid away and somehow directly provide the MLHandler as argument to the handler? We should discuss this in chat.

File: Monal/Classes/MLPromiseRegistry.m
Comment (tmolitor-stud-tu) [reply Monal/Classes/MLPromiseRegistry.m:69]: ah sorry!! I used the term `interface` (java jargon), but I meant `protocol` in objc terminology.

Comment (tmolitor-stud-tu) [reply Monal/Classes/MLPromiseRegistry.m:69]: the hierarchy was: `protocol` `MLPromise`, class implementing this protocol (by inheriting from this protocol): `MLSerializablePromise`. My fault, sorry!!

File: Monal/Classes/MLPromiseRegistry.h
Comment (tmolitor-stud-tu) [inline Monal/Classes/MLPromiseRegistry.h:28]: What's this `PromiseResolverCallback` for?

**Review by tmolitor-stud-tu — CHANGES_REQUESTED** [review]
I think we should chat about this.

**Comment by matthewrfennell** [comment]
I'm closing this PR in favour of https://github.com/monal-im/Monal/pull/1338. Once that one's merged, it should be easy to use the promises here too

**Pull request closed without merging**
